### ISSUE TITLE
Using a target size (torch.Size([64, 1])) that is different to the input size (torch.Size([304800, 1])) is deprecated. Please ensure they have the same size

### ISSUE THREAD

**Issue by KeerthiKrishna97** [bentrevett/pytorch-sentiment-analysis#105]
Hi, I am new to pytorch. There are 762 images in my dataset, each of size (1,160,160). I am trying to build a GAN and the code for Generator and Discriminator class is as follows:

batch_size = 64
sample_size = 64
nz = 128

class Generator(nn.Module):
    def __init__(self, nz):
        super(Generator, self).__init__()
        self.nz = nz
        self.main = nn.Sequential(
            nn.Flatten(),
            nn.Linear(self.nz, 256),
            nn.LeakyReLU(0.2),

            nn.Linear(256, 512),
            nn.LeakyReLU(0.2),

            nn.Linear(512, 1024),
            nn.LeakyReLU(0.2),

            nn.Linear(1024, batch_size),
            nn.Tanh(),
        )

    def forward(self, x):
        return self.main(x).view(-1, 1, 8, 8)

class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()
        self.n_input = batch_size
        self.main = nn.Sequential(
            nn.Flatten(),
            nn.Linear(self.n_input, 1024),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),

            nn.Linear(1024, 512),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),

            nn.Linear(512, 256),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),

            nn.Linear(256, 1),
            nn.Sigmoid(),
        )

    def forward(self, x):
        x = x.view(-1, batch_size)
        return self.main(x)

generator = Generator(nz).to(device)
discriminator = Discriminator().to(device)

But when I try to train it, it gives the following error:
_Using a target size (torch.Size([64, 1])) that is different to the input size (torch.Size([304800, 1])) is deprecated. Please ensure they have the same size_. 
I don't understand where the 304800 came from. Could anyone help me solve the issue?

**Comment by bentrevett** [comment]
Your error is in the line `nn.Linear(1024, batch_size)`.

I believe this should be `nn.Linear(1024, 8*8)` as the generator is trying to output 8x8 images.

**Comment by bentrevett** [comment]
Also, the line `x = x.view(-1, batch_size)` is an error. It should be `x = x.view(-1, batch_size)`.

**Comment by KeerthiKrishna97** [comment]
> Your error is in the line `nn.Linear(1024, batch_size)`.
> 
> I believe this should be `nn.Linear(1024, 8*8)` as the generator is trying to output 8x8 images.

Hello. Here 8*8 = 64 is the batch_size

**Comment by bentrevett** [comment]
> > Your error is in the line `nn.Linear(1024, batch_size)`.
> > I believe this should be `nn.Linear(1024, 8*8)` as the generator is trying to output 8x8 images.
> 
> Hello. Here 8*8 = 64 is the batch_size

Right, but `nn.Linear(1024, batch_size)` means it takes in a `[batch size, 1024]` tensor and outputs a `[batch size, batch size]` tensor, when it should be a `[batch size, channels * height * width]` tensor and then the last dimension should be reshaped.

**Comment by KeerthiKrishna97** [comment]
Okay. I have done that and now the error is :  mat1 and mat2 shapes cannot be multiplied (64x304800 and 64x1024)

**Comment by bentrevett** [comment]
Which line is giving you that error? Tensor shape mismatching is a common bug in deep learning models, and the best way to solve it by making sure the shapes output by each layer is what you expect by printing out the tensor shapes out by each layer and seeing if it matches with what you expect it to be.

**Comment by KeerthiKrishna97** [comment]
To tell you the line that's giving me this error, I will have to post the remaining part of the code as well. The continuation of the above code goes like this:

# to create real labels (1s)
def label_real(size):
    data = torch.ones(size, 1)
    return data.to(device)

# to create fake labels (0s)
def label_fake(size):
    data = torch.zeros(size, 1)
    return data.to(device)

# function to create the noise vector
def create_noise(sample_size, nz):
    return torch.randn(sample_size, nz).to(device)

def train_discriminator(optimizer, data_real, data_fake):
    #b_size = data_real.size(0)
    b_size = batch_size
    real_label = label_real(b_size)
    fake_label = label_fake(b_size)

    optimizer.zero_grad()

    output_real = discriminator(data_real)
    loss_real = criterion(output_real, real_label)

    output_fake = discriminator(data_fake)
    loss_fake = criterion(output_fake, fake_label)


    loss_real.backward()
    loss_fake.backward()
    optimizer.step()

    return loss_real + loss_fake
# function to train the generator network

def train_generator(optimizer, data_fake):
    #b_size = data_fake.size(0)
    b_size = batch_size
    real_label = label_real(b_size)

    optimizer.zero_grad()

    output = discriminator(data_fake)
    loss = criterion(output, real_label)

    loss.backward()
    optimizer.step()

    return loss 

# create the noise vector
noise = create_noise(sample_size, nz)

generator.train()
discriminator.train()

for epoch in range(epochs):
    loss_g = 0.0
    loss_d = 0.0
    for bi, x_train in tqdm(enumerate(train_loader), total=int(len(image_datasets[0])/train_loader.batch_size)):
        image = image_datasets[0]
        image = torch.Tensor(image)
        image = image.to(device)
        b_size = len(image)
        # run the discriminator for k number of steps
        for step in range(k):
            data_fake = generator(create_noise(b_size, nz)).detach()
            data_real = image
            # train the discriminator network
            loss_d += train_discriminator(optim_d, data_real, data_fake)
        data_fake = generator(create_noise(b_size, nz))
        # train the generator network
        loss_g += train_generator(optim_g, data_fake)

The error is in the line:
 loss_d += train_discriminator(optim_d, data_real, data_fake)
and the optimizer used is Adam

**Comment by KeerthiKrishna97** [comment]
> Your error is in the line `nn.Linear(1024, batch_size)`.
> 
> I believe this should be `nn.Linear(1024, 8*8)` as the generator is trying to output 8x8 images.

I believe the error is here. The output of nn.Linear of Generator shouldn't be batch size. As you said, it should be channels × height× width
